src/compiler/moduleSpecifiers.ts

```diff
@@ -1036,26 +1023,46 @@ function getPathsRelativeToRootDirs(path: string, rootDirs: readonly string[], g
     });
 }
 
-function removeExtensionAndIndexPostFix(fileName: string, ending: Ending, options: CompilerOptions, host?: ModuleSpecifierResolutionHost): string {
-    if (fileExtensionIsOneOf(fileName, [Extension.Json, Extension.Mjs, Extension.Cjs])) return fileName;
+function processEnding(fileName: string, allowedEndings: readonly ModuleSpecifierEnding[], options: CompilerOptions, host?: ModuleSpecifierResolutionHost): string {
+    if (fileExtensionIsOneOf(fileName, [Extension.Json, Extension.Mjs, Extension.Cjs])) {
+        return fileName;
+    }
+
     const noExtension = removeFileExtension(fileName);
-    if (fileName === noExtension) return fileName;
-    if (fileExtensionIsOneOf(fileName, [Extension.Dmts, Extension.Mts, Extension.Dcts, Extension.Cts])) return noExtension + getJSExtensionForFile(fileName, options);
-    switch (ending) {
-        case Ending.Minimal:
+    if (fileName === noExtension) {
+        return fileName;
+    }
+
+    if (fileExtensionIsOneOf(fileName, [Extension.Dmts, Extension.Mts, Extension.Dcts, Extension.Cts])) {
+        return noExtension + getJSExtensionForFile(fileName, options);
+    }
+
+    switch (allowedEndings[0]) {

```

Comment:
I don't understand why looking only at the first allowedEnding works. Is it because there are only a small number of sets, and the first one of each is known?

Reply:
I almost made this function recursive with the tail of this array, but that would have resulted in some duplicated work and it turns out that it’s only rare circumstances that  we can’t respect the first choice (these are sorted in preferrerrerrerred order). Under `case ModuleSpecifierEnding.TsExtension`, we find the next suitable alternative with `findIndex`. Under `ModuleSpecifierEnding.Minimal`, we actually don’t bother looking at the order of the rest of the array, because it assumes that if you can’t get a really minimal ending, you probably want the next-most-minimal ending. This function is a little weird because not all imaginable `allowedEndings` arrays are actually realistic, and some of those assumptions are hard-coded such that you could call it with a nonsense `allowedEndings` array and find that it was simply not respected. The function would need to be reworked a little bit if it needed to respect `allowedEndings` 100%.